modules/distributor/queue/queue.go

```diff
@@ -0,0 +1,224 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/common/user"
+	"go.uber.org/atomic"
+)
+
+type request[T any] struct {
+	data T
+}
+
+type ProcessFunc[T any] func(ctx context.Context, data T) error
+
+// Queue represents a single tenant's queue.
+type Queue[T any] struct {
+	// wg is used to wait for the workers to drain the queue while stopping
+	wg sync.WaitGroup
+
+	logger log.Logger
+
+	name           string
+	tenantID       string
+	workerCount    int
+	size           int
+	reqChan        chan *request[T]
+	fn             ProcessFunc[T]
+	workersCloseCh chan struct{}
+
+	pushesTotalMetrics         *prometheus.CounterVec
+	pushesFailuresTotalMetrics *prometheus.CounterVec
+	lengthMetric               *prometheus.GaugeVec
+
+	readOnly *atomic.Bool
+}
+
+func New[T any](cfg Config, logger log.Logger, reg prometheus.Registerer, fn ProcessFunc[T]) *Queue[T] {
+	pushesTotalMetrics := prometheus.NewCounterVec(prometheus.CounterOpts{
```

Comment:
we already have a lot of metrics here. i think we either need to keep these and remove the external metrics that are basically the same thing or expect individual users of this queue to handle their own metrics.

if we do remove the metrics in forwader.go in favor of these we need to list a breaking change in the changelog.

Reply:
I would prefer that we start using these metrics in the queue and remove the old forwarder metrics that are the exact same thing. However, this would require changes to our alerts and runbooks due to renaming these two metrics.

In order to move this forward let's do this:
- Keep both sets of metrics.
- Mark the old ones deprecated 
  - add a comment in code that they are deprecated
  - add it to the prom metric descrption
  - call out the deprecation in the changelog and indicate the new metric.
- In a future change we can remove the old ones and update all of our related operational code.

Reply:
Done in [f995cac](https://github.com/grafana/tempo/pull/1796/commits/f995cace7f4a12e4b34f81f25e17dec60d456508).